Add list of DateTimeFieldSpec to Schema

npawar opened the pull request:
Introduce DateTimeFieldSpec to pinot Schema for the following reasons:
1. Have an ability in the pinot schema to distinguish between time column format and data granularity. Currently there's no way to specify data granularity, and the timeSpec is being used stretched to do both in one shot.
2. Bring in a placeholder for creating a uniform time column across all pinot tables (will be useful in doing rollups)
3. Provision for supporting multiple time columns of different type (PRIMARY, DERIVED etc).

mcvsubbu commented on pinot-common/src/main/java/com/linkedin/pinot/common/data/DateTimeFieldSpec.java:
```diff
@@ -0,0 +1,222 @@
+    Preconditions.checkArgument(formatTokens.length == 3 || formatTokens.length == 4,
+        "format must be of format size:timeunit:timeformat(:pattern)");
+    Preconditions.checkArgument(formatTokens[0].matches(NUMBER_REGEX)
+        && EnumUtils.isValidEnum(TimeUnit.class, formatTokens[1]), "format must be of format [0-9]+:<TimeUnit>:<TimeFormat>(:pattern)");
```
private static string ERROR_STR = "format must be ...."

npawar replied:
Done

mcvsubbu commented on pinot-common/src/main/java/com/linkedin/pinot/common/data/DateTimeFieldSpec.java:
```diff
@@ -0,0 +1,222 @@
+@JsonIgnoreProperties(ignoreUnknown = true)
+public final class DateTimeFieldSpec extends FieldSpec {
+
+  private static final String NUMBER_REGEX = "[0-9]+";
```
is 0 allowed? what does 0 mean?

npawar replied:
Fixed to not allow 0

mcvsubbu commented on pinot-common/src/main/java/com/linkedin/pinot/common/data/DateTimeFieldSpec.java:
```diff
@@ -0,0 +1,222 @@
+    return _dateTimeType;
+  }
+
+  public void setDateTimeType(DateTimeType dateTimeType) {
```
check which of these setters we actually need. If we can avoid individual setters that is preferred, but I suspect the json parsers need the setters? At the minimum add an annottion & javadoc so that nobody uses the setters. Perhaps @VisibleForTesting will work?

npawar replied:
Tried to follow what was done in the rest of the setters of the schema, which is either a comment saying "// Required by JSON de-serializer. DO NOT REMOVE." or no comment/annotation at all. Not sure if  @VisibleForTesting  makes sense here. Will add comments/javadocs

npawar replied:
@VisibleForTesting does not seem to disallow usage, even from IDE. Added javadoc, stating these should not be used.

mcvsubbu commented on pinot-common/src/main/java/com/linkedin/pinot/common/data/Schema.java:
```diff
@@ -121,6 +130,19 @@ public void setMetricFieldSpecs(@Nonnull List<MetricFieldSpec> metricFieldSpecs)
+    return _dateTimeFieldSpecs;
+  }
+
+  public void setDateTimeFieldSpecs(@Nonnull List<DateTimeFieldSpec> dateTimeFieldSpecs) {
```
re-checking whether this setter is needed. Same comment as before in terms of annotation and javadoc comments.

npawar replied:
I tried to make it consistent with the already existing metricFieldSpecs and dimensionFieldSpecs and timeFieldSpec. Those have setters, with no annotations. Should I add the annotations to those as well then?

mcvsubbu commented on pinot-common/src/main/java/com/linkedin/pinot/common/data/Schema.java:
```diff
@@ -121,6 +130,19 @@ public void setMetricFieldSpecs(@Nonnull List<MetricFieldSpec> metricFieldSpecs)
+    Preconditions.checkState(_dateTimeFieldSpecs.isEmpty());
+
+    for (DateTimeFieldSpec dateTimeFieldSpec : dateTimeFieldSpecs) {
+      addField(dateTimeFieldSpec);
```
Do you need to add it to the set as well as to the list?

npawar replied:
I tried to make it consistent with the already existing metricFieldSpecs and dimensionFieldSpecs. Those have sets and lists, and methods to add into both. Not sure why it was done for those, and not sure if they will be used for dateTimeFieldSpec, but wanted to have them there to be consistent with how it is designed.

mcvsubbu commented on pinot-common/src/main/java/com/linkedin/pinot/common/data/Schema.java:
```diff
@@ -288,6 +334,7 @@ public boolean validate(Logger ctxLogger) {
         switch (fieldType) {
           case DIMENSION:
           case TIME:
+          case DATE_TIME:
```
can fieldType be a string/float/double?

npawar replied:
No, it's an enum consisting of only these values which are in the cases (DIMENSION, TIME, DATE_TIME, METRIC)

mcvsubbu replied:
I was referring to the switch below, that checks for different dataTypes.

npawar replied:
Yes, it can be string in case of more advanced simple date formats. And since it is an n units since epoch calculation, better to not restrict to whole numbers

mcvsubbu commented on pinot-common/src/main/java/com/linkedin/pinot/common/data/Schema.java:
```diff
@@ -475,6 +522,31 @@ public SchemaBuilder addTime(@Nonnull TimeGranularitySpec incomingTimeGranularit
       return this;
     }
 
+    public SchemaBuilder addDateTime(@Nonnull String name, @Nonnull DataType dataType, @Nonnull String format,
```
Not sure if we need so many different versions of addDateTime. Start with one, and add more as needed.

npawar replied:
Cleaned up addDateTime methods clutter, keeping just 1

kishoreg replied:
can we take DateTimeFieldSpec instead of individual each value. One reason why we had soo many methods for addTime was every time we enhanced TimeFieldSpec we added methods for every possible combination.

mcvsubbu commented on pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java:
```diff
@@ -31,7 +30,7 @@
  * <p>- <code>TimeType</code>: time unit of the time column (e.g. MINUTES, HOURS).
  * <p>- <code>TimeUnitSize</code>: size of the time buckets (e.g. 10 MINUTES, 2 HOURS). By default this is set to 1.
- * <p>- <code>TimeFormat</code>: Can be either EPOCH (default) or SIMPLE_DATE_FORMAT:pattern e.g SIMPLE_DATE_FORMAT:yyyyMMdd 
+ * <p>- <code>TimeFormat</code>: Can be either EPOCH (default) or SIMPLE_DATE_FORMAT:pattern e.g SIMPLE_DATE_FORMAT:yyyyMMdd
```
If u did not change this file at all (and it seems like that), then please restore the master version of the file

npawar replied:
I removed an import, the IDE seems to have added all the other line formatting.

mcvsubbu commented on pinot-common/src/test/java/com/linkedin/pinot/common/data/FieldSpecTest.java:
```diff
@@ -192,6 +195,69 @@ public void testTimeFieldSpecConstructor() {
+   * Test {@link TimeFieldSpec} constructors.
+   */
+  @Test
+  public void testDateTimeFieldSpecConstructor() {
```
I would add negative test cases as well, with 0, -1, etc for numbers

npawar replied:
Refactored the tests to address the comments related to tests

mcvsubbu commented on pinot-common/src/test/java/com/linkedin/pinot/common/data/FieldSpecTest.java:
```diff
@@ -192,6 +195,69 @@ public void testTimeFieldSpecConstructor() {
+
+  }
+
+  @Test(expectedExceptions = {IllegalArgumentException.class})
```
If you get IllegalArgumentException in line 233, the other lines will not be executed, right? Isnt it better to actually test for the exception in the line in which you expect it to happen?

npawar replied:
Fixed

npawar commented on pinot-common/src/main/java/com/linkedin/pinot/common/data/DateTimeFieldSpec.java:
```diff
@@ -0,0 +1,222 @@
+  /**
+   * @param name
+   * @param dataType
+   * @param format - size:timeunit:timeformat eg: 1:MILLISECONDS:EPOCH, 5:MINUTES:EPOCH, 1:DAYS:SIMPLE_DATE_FORMAT:yyyyMMdd
```
Done

npawar commented on pinot-common/src/main/java/com/linkedin/pinot/common/data/DateTimeFieldSpec.java:
```diff
@@ -0,0 +1,222 @@
+  private DateTimeType _dateTimeType;
+
+  public enum DateTimeType {
+    PRIMARY,
```
Done

npawar commented on pinot-common/src/main/java/com/linkedin/pinot/common/data/DateTimeFieldSpec.java:
```diff
@@ -0,0 +1,222 @@
+   * @param defaultNullValue
+   */
+  public DateTimeFieldSpec(@Nonnull String name, @Nonnull DataType dataType, @Nonnull String format,
+      @Nonnull String granularity, DateTimeType dateTimeType, Object defaultNullValue) {
```
Cleaned up constructors clutter, keeping only a single one which will be used immediately.

npawar commented:
@mcvsubbu I have added the design doc on the wiki page in the Design Docs section: https://github.com/linkedin/pinot/wiki/Schema-TimeSpec-Refactoring
Addressed all your comments as well, please review again!

kishoreg commented on pinot-common/src/main/java/com/linkedin/pinot/common/data/DateTimeFieldSpec.java:
```diff
@@ -0,0 +1,222 @@
+  private String _granularity;
+  private DateTimeType _dateTimeType;
+
+  public enum DateTimeType {
```
I would like to have a method in Pinot that takes DateTimeSpec and a value and returns a DateTime object. This can be util or part of this class

npawar replied:
Will add this in another PR

kishoreg commented:
with this change the incomingTimeGranularity will be removed from Schema but this is needed by the segment generation/realtime ingestion code. We need to make sure we add this in tableConfig.
